formats/mdm/MDM_2_4.ddl

```diff
@@ -0,0 +1,515 @@
+    --
+    -- See 6.1.2.1.2: MDM Acknowledge
+    -- See 6.2.3: MDM Acknowledgement Message (Type 2) Specification
+    let raw_ack_byte = UInt8
+
+    -- See 6.1.2.1.3: Version
+    version = MDM_Version
+
+    -- See 6.1.2.1.4: MDM Sequence number
+    sequence_number = BEUInt16
+
+    -- See 6.1.2.1.5: ML2B interface ID
+    ml2b_interface_id = ML2B_Interface_id
+
+    -- See 6.1.2.1.6: MDM type
+    type = MDM_Type raw_ack_byte
+
+-- See 6.1.2.1.3: Version, major followed by minor; e.g. 2.4 yields
+-- major version 2, minor version 4.
+def MDM_Version =
+  block
+    major = UInt8
+    minor = UInt8
+
+-- See 6.1.2.1.5: ML2B interface ID
+def ML2B_Interface_id =
+  block
+    origination_id = BEUInt16
+    destination_id = BEUInt16
+
+-- Parse the message type followed by the number of messages field. This
+-- returns only the type since the number of messages is always implied
+-- by the type anyway.
+--
+-- See:
+-- * 6.1.2.1.6: MDM type
+-- * 6.1.2.1.7: Number of messages
+def MDM_Type (raw_ack_byte: uint 8) =
+  First
+    -- Parse the message type octet followed by the number of messages
+    -- octet, which has a fixed value corresponding to each message
+    -- type.
+    Type_VRT                  = @Match [0x0, 0x1, 0x0, 0x1]
+    Type_Time_Of_Day          = @Match [0x0, 0x3, 0x0, 0x1]
+    Type_Signal_Port_User_ID  = @Match [0x0, 0x4, 0x0, 0x1]
+    Type_Health_Status        = @Match [0x0, 0x5, 0x0, 0x1]
+    Type_Command              = @Match [0x0, 0x6, 0x0, 0x1]
+    Type_Switch_Group_User_ID = @Match [0x0, 0x7, 0x0, 0x1]
+
+    Type_Acknowledgement = block
+      @Match [0x0, 0x2, 0x0, 0x0]
+      -- See 6.3.2: MDM Acknowledgement Message (Type 2) Specification
+      --
+      -- In the Acknowledge messge type, the "ack" octet parsed in the
+      -- message header is treated as a response octet which takes one
+      -- of a number of meaningful values.
+      --
+      -- The specification states that all other possible values are
+      -- Reserved for Future Use; if they are encountered here, they
+      -- result in a parse failure.
+      response = raw_ack_byte as? MDM_Ack_Response_Value
+
+-- See 6.3.2: MDM Acknowledgement Message (Type 2) Specification, Table
+-- 74
+bitdata MDM_Ack_Response_Value where
+  -- Acknowledged (A)
+  Resp_A = 0x01
+  -- Invalid Message (IM)
+  Resp_IM = 0x02
+  -- Invalid Function (IF)
+  Resp_IF = 0x04
+  -- Invalid Range (IR)
+  Resp_IR = 0x08
+  -- Resource Unavailable (RU)
+  Resp_RU = 0x10
+  -- Unexpected Sequence Number (US)
+  Resp_US = 0x20
+  -- Unauthorized User (UU)
+  Resp_UU = 0x40
+
+-- See 6.1.2.2, MDM Body
+def MDM_Body (ty: MDM_Type) =
+  case ty of
+    Type_VRT ->
+      -- See 6.3.1.2.1: VRT Message Format
+      -- VRT_Message
+      Fail "VRT message parsing not implemented"
+
+    -- See 6.3.2: MDM Acknowledgement Message (Type 2) Specification
+    -- Acknowledgement messages have no body.
+    Type_Acknowledgement info ->
+      ^ {| Body_Acknowledgement = { } |}
```

Comment:
Probably skip the ^
Also, I think just writing `{| Body_Acknowledgement |}` should work.

Reply:
Fixed in c63542ff1b603cf2a9576ff2169d05e86088f4e5